### ISSUE TITLE
Codex packet: static strategic curriculum masks for PPO manager targets

### ISSUE THREAD

**Issue by BillXu21** [BillXu21/Kaggriculture#38]
## Goal

Add a small opt-in **training curriculum restriction** to the existing daily-manager PPO action distribution so experiments can temporarily restrict:

- maximum land target;
- maximum GOOSE target;
- maximum COW target;
- maximum SHEEP target.

This is curriculum scaffolding for scratch/self-play experiments. It is **not** a permanent strategy heuristic, mechanical legality mask, reward change, executor change, or automatic curriculum scheduler.

The immediate research question is whether reducing the manager's strategic search space can make scratch PPO improve more consistently without relying on manual elite-trajectory selection.

## Base / source-of-truth

Use the current Stage-2 training worktree supplied for this task.

Before editing:

1. read repository `AGENTS.md`;
2. report the exact branch and HEAD SHA;
3. inspect the current versions of the relevant PPO policy/update/adapter/checkpoint/CLI source and focused tests;
4. verify that the worktree contains the current economic PPO path used by the operator, including `terminal_own_bank` / current-current economic training where applicable.

Do not assume `main` or an older issue branch is current merely from historical documentation.

If the supplied worktree does not contain the current Stage-2 PPO path, STOP and report the mismatch rather than implementing against a stale base.

Do not rebase or force-push shared history.

## Current architectural boundary

Preserve:

`daily manager -> deterministic executor -> primitive actions`

The curriculum restriction applies only to the learned manager's **strategic target distributions**.

For this packet:

- animal feeding remains deterministic executor maintenance;
- executor placement/routing/lifecycle mechanics remain unchanged;
- CARE behavior remains unchanged;
- fertilizer behavior remains unchanged;
- crop-target distributions remain unchanged;
- sell presence/quantity behavior remains unchanged;
- opening behavior remains unchanged;
- terminal reward behavior remains unchanged.

Do not edit `executor_v0/**` except for a test-only import seam if absolutely unavoidable.

### 1. Add a small immutable curriculum config

Add a frozen/hashable config with semantics equivalent to:

```python
@dataclass(frozen=True)
class CurriculumMaskConfig:
    max_land: int | None = None
    max_goose: int | None = None
    max_cow: int | None = None
    max_sheep: int | None = None
```

Exact naming may differ if current source suggests a better local convention.

Semantics:

- `None` means unrestricted/current behavior;
- `max_land` is in decoded land-target units `1..4`;
- animal maxima are absolute manager target counts in `0..count_max`;
- `max_goose=0` means the GOOSE target categorical may choose only class 0;
- no field may silently clamp an invalid value; invalid config fails loudly.

Use named species fields or another serialization that cannot silently depend on an undocumented animal-order tuple.

Do **not** add crop/fertilizer/CARE/sell masks in this packet.

### 2. Mask the actual PPO distribution, never post-process plans

Add one pure helper equivalent to:

```python
apply_curriculum_mask(
    logits,
    curriculum_config,
) -> masked_logits
```

For categorical classes outside the configured support, assign zero probability using a numerically safe masked-logit representation.

At least one class must remain valid for every categorical component.

The mask must be applied before:

- stochastic rollout sampling;
- deterministic/greedy action decode;
- rollout log-prob calculation;
- entropy calculation;
- `evaluate_actions()` stored-action log-prob recomputation;
- PPO minibatch/current-policy log-prob recomputation.

Do **not** sample an unrestricted action and overwrite the resulting `DailyPlan`. The action executed by the environment must be the same action whose probability PPO records.

### 3. PPO update must use the exact same curriculum support as rollout

Thread the same `CurriculumMaskConfig` through the PPO policy and PPO update path.

A trajectory collected under mask A must never be updated as though it were collected under mask B.

Before a PPO update or explicit stored-action evaluation, fail loudly if any stored action is outside the active curriculum support.

Examples:

- active `max_land=1`, stored land action `2` -> fail;
- active `max_goose=0`, stored goose target `1` -> fail.

Do not allow `-inf` stored-action log-probs to flow into ratio/KL math and create NaNs as a substitute for validation.

### 4. KL-to-frozen semantics

If `kl_to_frozen_coef == 0`, preserve existing behavior except for the active curriculum distribution.

If KL-to-frozen is enabled, compare current and frozen policies on the **same curriculum-restricted support**:

- apply the same curriculum mask to current logits;
- apply the same curriculum mask to frozen logits;
- then compute the existing analytic KL.

This makes KL measure drift relative to frozen E conditional on the action space currently available to the curriculum.

Do not modify frozen parameters.

### 5. Policy identity must include curriculum semantics

Two policies with identical parameter tensors but different curriculum masks are different behavior distributions.

Update PPO policy/adapter identity construction so the effective identity/fingerprint includes:

- parameter fingerprint;
- canonical curriculum-mask serialization/version.

Requirements:

- same params + same mask -> same identity;
- same params + different mask -> different identity;
- unrestricted/default mask remains deterministic;
- do not change unrelated deterministic BC policy identities.

This is required so trajectory provenance/on-policy routing cannot mix curriculum stages silently.

### 6. CLI / experiment plumbing

Add small optional CLI/config plumbing to the current training/evaluation entry point.

Equivalent flags are acceptable:

```text
--curriculum-max-land N
--curriculum-max-goose N
--curriculum-max-cow N
--curriculum-max-sheep N
```

Omitting all flags must reproduce current unrestricted behavior.

Do not add a matrix runner or automatic scheduler in this packet.

A later run must be able to load the same PPO train state/checkpoint and explicitly use a different curriculum mask for the next stage.

Changing the mask is allowed only at a clear run/update boundary, never halfway through a pending rollout/update.

### 7. Provenance

Record the active curriculum config in run/checkpoint provenance using the repository's current provenance mechanism.

Do not gratuitously break the PPO checkpoint format.

Old checkpoints with no curriculum metadata must load as:

```text
curriculum = unrestricted
```

When a checkpoint is resumed under a different explicit mask, preserve enough provenance to distinguish:

- parent checkpoint's recorded curriculum;
- current run's active curriculum.

The active mask printed in logs/artifacts must be explicit rather than inferred from filenames.

### 8. No-mask compatibility is a hard invariant

With the unrestricted/default curriculum:

- stochastic actions for the same params/input/RNG must match pre-change behavior;
- deterministic actions must match pre-change behavior;
- log-probs must match;
- entropy must match;
- PPO update behavior must match within the existing deterministic/numerical contract.

Do not accept a default behavior change merely because the new path is mathematically equivalent in theory.

## Required focused tests

Add focused tests covering at minimum:

### Unrestricted compatibility

- default/no curriculum mask produces the same deterministic action tensors as the old path;
- same fixed RNG produces the same stochastic action tensors as the old path;
- same log-prob and entropy outputs under the existing tolerance.

### Land restriction

For `max_land=1`:

- stochastic sampling always yields decoded land target 1;
- deterministic decode always yields land target 1;
- stored-action log-prob recomputation matches the rollout value;
- stored land target >1 is rejected loudly.

Also test a nontrivial cap such as `max_land=2`.

### Animal restriction

For `max_goose=0`:

- stochastic and deterministic goose target is always 0;
- COW and SHEEP distributions remain otherwise unchanged;
- stored goose target >0 is rejected loudly.

Test at least one nonzero cap, e.g. `max_goose=2`, proving classes above 2 cannot be sampled.

### Multiple restrictions

A combined land + per-species animal mask must work without changing crop, fertilizer, CARE, or sell action tensors merely due to mask plumbing.

### PPO update

Collect/evaluate a small synthetic or existing trajectory under a mask and prove:

- rollout stored log-prob equals recomputed pre-update log-prob;
- one PPO update completes with finite metrics;
- no masked-out action enters the update;
- unrestricted PPO update remains backwards compatible.

### KL

With nonzero KL-to-frozen coefficient under a restrictive mask:

- KL is finite;
- both distributions use identical restricted support.

### Identity

- same params/unrestricted mask has stable identity;
- same params + `max_goose=0` differs;
- same params + `max_land=1` differs;
- rebuilding the same mask produces the same identity.

### Checkpoint/provenance

- old checkpoint metadata without curriculum loads as unrestricted;
- new checkpoint/run provenance records the exact active mask;
- the same loaded train state can be explicitly reconstructed under a different curriculum mask for a later curriculum stage without mutating the stored parameters.

## Small real-path smoke

After focused tests, run one small existing fast-engine PPO smoke using an intentionally obvious restriction such as:

```text
max_land=1
max_goose=0
```

The purpose is plumbing only.

Verify:

- rollout completes;
- every trainable row respects the restriction;
- stored/recomputed log-probs agree according to the existing PPO invariant;
- one PPO update completes with finite metrics;
- checkpoint/result provenance contains the exact mask;
- no executor/runtime/fallback regression appears.

Do not run a long training experiment in this implementation packet.

## Non-goals

- no reward changes;
- no reward scaling;
- no short-horizon/cutoff training;
- no no-opening training;
- no automatic curriculum advancement;
- no performance threshold for unmasking;
- no state-dependent masks;
- no crop-target masks;
- no fertilizer masks;
- no CARE masks;
- no sell masks;
- no mechanical legality-mask framework;
- no model architecture change;
- no executor strategy or heuristic change;
- no opening change;
- no E-history semantic change;
- no elite-trace BC/self-imitation;
- no promotion-ratchet redesign;
- no hyperparameter sweep;
- no TPU long run.

## Acceptance

This packet is complete when:

1. current PPO can be run unrestricted with no behavioral regression;
2. land and per-species animal target supports can be restricted through explicit config;
3. sampling, greedy decode, entropy, rollout log-prob, stored-action recomputation, and PPO updates all use the same restricted distribution;
4. out-of-support trajectory actions fail loudly;
5. KL-to-frozen remains well-defined under masking;
6. policy identity distinguishes identical weights under different curriculum masks;
7. old checkpoints remain usable and new runs record exact mask provenance;
8. one bounded real-path smoke passes.

## Completion report

Report:

- starting branch and SHA;
- final branch and SHA;
- exact files changed;
- final curriculum config/API;
- how policy identity incorporates the mask;
- how masked logits are shared between rollout and PPO update;
- old-checkpoint compatibility behavior;
- focused tests and results;
- real-path smoke command/result;
- one example unrestricted config;
- one example `max_land=1, max_goose=0` config;
- any current-source difference that required deviating from this packet.

STOP after the static curriculum-mask seam and bounded smoke are complete.

Do not design tonight's experimental matrix, short-horizon curriculum, no-opening training, or elite-trace consolidation in this packet.
